cfme/cloud/provider/__init__.py

```diff
@@ -123,7 +126,77 @@ def discover_dict(credential):
         raise NotImplementedError("This provider doesn't support discovery")
 
 
+@attr.s
+class CloudProviderCollection(BaseCollection):
+    """Collection object for CloudProvider object
+    """
+
+    ENTITY = CloudProvider
+
+    def all(self):
+        view = navigate_to(self, 'All')
+        provs = view.entities.get_all(surf_pages=True)
+
+        # trying to figure out provider type and class
+        # todo: move to all providers collection later
+        def _get_class(pid):
+            prov_type = self.appliance.rest_api.collections.providers.get(id=pid)['type']
+            for prov_class in provider_types('cloud').values():
+                if prov_class.db_types[0] in prov_type:
+                    return prov_class
+
+        return [self.instantiate(prov_class=_get_class(p.data['id']), name=p.name) for p in provs]
+
+    # todo: think over, need to get rid of this definition
+    def instantiate(self, prov_class, *args, **kwargs):
+        return prov_class.from_collection(self, *args, **kwargs)
+
+    def create(self, prov_class, *args, **kwargs):
+        # ugly workaround until I move everything to main class
+        class_attrs = [at.name for at in attr.fields(prov_class)]
+        init_kwargs = {}
+        create_kwargs = {}
+        for name, value in kwargs.items():
+            if name not in class_attrs:
+                create_kwargs[name] = value
+            else:
+                init_kwargs[name] = value
+
+        obj = self.instantiate(prov_class, *args, **init_kwargs)
+        obj.create(**create_kwargs)
```

Comment:
Required: please return provider instance here, or if it not needed here please explain why we don't need it here?

Reply:
fixed